src/Common/noexcept_scope.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+#include <base/scope_guard.h>
+#include <Common/Exception.h>
+#include <Common/LockMemoryExceptionInThread.h>
+
+
+#define NOEXCEPT_SCOPE_IMPL_CONCAT(n, expected) \
+    LockMemoryExceptionInThread lock_memory_tracker##n(VariableContext::Global);   \
+    SCOPE_EXIT(                                                                    \
+        {                                                                          \
+            const auto uncaught = std::uncaught_exceptions();                      \
+            assert((expected) == uncaught || (expected) + 1 == uncaught);          \
+            if ((expected) < uncaught)                                             \
+            {                                                                      \
+                tryLogCurrentException("NOEXCEPT_SCOPE");                          \
```

Comment:
This will not work, you cannot rethrow exception outside the catch block, this will simply terminate (via std::terminate) the program. So this `NOEXCEPT_SCOPE` macro will simply call `std::terminate` on exception and will lost original exception.

I was thinking about this a little, and the only thing that pops up in my mind is setting some flag here and check it in Exception ctor and if it set, call abort/std::terminate, that way you will not loose it, however it looks not great.

Reply:
You're right, it does not work as expected and seems like it's there's no way to get exception that is currently unwinding stack (even using `__cxa_*` functions or something from libunwind...). Not sure what options is better: just remove this like or add thread-local flag and check it when throwing an exception. Maybe it's ok to simply remove it for now and think about logging if we will face some issue that we will not be able to debug without information about unexpected exception...

Reply:
Here is an RFC - #39229